### PR TITLE
Improve chip erase emulation for dryboot/dryrun

### LINKED ISSUE #1833: Terminal erase command
Here's a thought. The `erase` command is very useful, and I often use it to _just_ wipe the EEPROM (`erase eeprom`) or flash (`erase flash`). However, `erase` alone performs a chip erase, and depending on the fuse setting, the EEPROM is either preserved or erased. This may comae as a surprise to some, as no extra argument is provided.

Wouldn't it be better if the "standard" chip erase command was something like `erase chip`, and instead print the help text when the user enter `erase`?

```
avrdude: processing -t interactive terminal
avrdude> erase 
erasing chip ...
avrdude> erase flash
Caching | ################################################## | 100% 0.01 s 
avrdude> erase eeprom
Caching | ################################################## | 100% 0.00 s 
avrdude> flush
avrdude: synching cache to device ... 
Writing | ################################################## | 100% 0.00 s 
```

### PR DESCRIPTION
**Pull request by stefanrueger**
Fixes #1833

### CODE REVIEW

File: src/term.c [hunk 1]
```diff
@@ -964,7 +964,16 @@ static int cmd_erase(const PROGRAMMER *pgm, const AVRPART *p, int argc, const ch
     return cmd_write(pgm, p, 6, args);
   }
 
-  term_out("erasing chip ...\n");
+  const char *note = "";
+  if(avr_locate_eeprom(p)) {
+    int eesave, bakverb = verbose;
+    verbose = -123;             // Silently read EESAVE configuration bit
+    if(avr_get_config_value(pgm, p, "eesave", &eesave) == 0)
```
Comment (dl8dtl) [inline src/term.c:971]: What happens for (ancient) devices that don't have this configuration value?

Comment (stefanrueger) [reply src/term.c:971]: Here are those parts *with* EEPROM but *without* (known) `eesave` configuration

```
for p in $(avrdude -p*/St | grep eeprom.size | ct -f2); do
  echo $p $(avrdude -c dryrun -qqp $p -T "con eesave" 2>&1)
done |& grep -P "(error|warning)"
```

They fall into two categories:
 - Ancient device that truly doesn't have the configuration: `ATtiny12`, `ATtiny15`, `AT90S1200`, `AT90S4414`, `AT90S2313`, `AT90S2333`, `AT90S2343`, `AT90S2323`, `ATtiny22`, `AT90S4433`, `AT90S8515`, `AT90S8535`, `AT90S4434`, `ATmega163`, `ATmega161`
 - No ATDF file and/or no data sheet for neither love nor money (Part never existed? some knockoff part with emulated EEPROM?): `ATxmega64A4`, `ATxmega128A4`, `ATxmega192A1`, `ATxmega256A1`, `AVR8EA28`, `AVR8EA32`, `LGT8F88P`, `LGT8F168P`, `LGT8F328P`

Either way, the PR will say erasing chip ... without asserting anything about EEPROM, ie, neither claims erasing chip (including EEPROM) ... nor claims erasing chip (but not EEPROM) ...

@dl8dtl But what *do* these parts to on a CE? Erase EEPROM?

Comment (stefanrueger) [reply src/term.c:971]: $ avrdude -qqp t22 -c dryrun -T erase
erasing chip ...

$avrdude -qqp t26 -c dryrun -T erase
erasing chip (including EEPROM) ...
```

Comment (stefanrueger) [reply src/term.c:971]: ... but I just realised that `erase` shouldn't affirm whether or not EEPROM is erased if the programmer is a bootloader; what the bootloader does with EEPROM on a CE is anyone's guess. So fixed *that*.

**Comment by stefanrueger** [comment]
@MCUdude Would you like to test this PR?

**Comment by MCUdude** [comment]
Thanks for the PR @stefanrueger!

It works perfectly fine on the ATmega8535 and ATtiny9, but on the AVR64DU28 (using a SerialUPDI programmer) it erases the EEPROM regardless... 

(BTW we, the family, are going on vacation for a few weeks, so don't expect much from me. I'll read emails and comments on Github though.)

```
$ ./avrdude -cserialupdi -p64du28 -P /dev/cu.usbserial-1410 -t
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9622 (probably 64du28)

avrdude> write flash 0 0x300 "Hello World" ...
Caching | ################################################## | 100% 0.00 s 
avrdude> write eeprom 0 -1 "eep " ...
Caching | ################################################## | 100% 0.00 s 
avrdude> flush
avrdude: synching cache to device ... 
Writing | ################################################## | 100% 9.82 s 
avrdude> read flash
Reading | ################################################## | 100% 0.00 s 
0100  6f 20 57 6f 72 6c 64 00  48 65 6c 6c 6f 20 57 6f  |o World.Hello Wo|
0110  72 6c 64 00 48 65 6c 6c  6f 20 57 6f 72 6c 64 00  |rld.Hello World.|
0120  48 65 6c 6c 6f 20 57 6f  72 6c 64 00 48 65 6c 6c  |Hello World.Hell|
0130  6f 20 57 6f 72 6c 64 00  48 65 6c 6c 6f 20 57 6f  |o World.Hello Wo|
0140  72 6c 64 00 48 65 6c 6c  6f 20 57 6f 72 6c 64 00  |rld.Hello World.|
0150  48 65 6c 6c 6f 20 57 6f  72 6c 64 00 48 65 6c 6c  |Hello World.Hell|
0160  6f 20 57 6f 72 6c 64 00  48 65 6c 6c 6f 20 57 6f  |o World.Hello Wo|
0170  72 6c 64 00 48 65 6c 6c  6f 20 57 6f 72 6c 64 00  |rld.Hello World.|
0180  48 65 6c 6c 6f 20 57 6f  72 6c 64 00 48 65 6c 6c  |Hello World.Hell|
0190  6f 20 57 6f 72 6c 64 00  48 65 6c 6c 6f 20 57 6f  |o World.Hello Wo|
01a0  72 6c 64 00 48 65 6c 6c  6f 20 57 6f 72 6c 64 00  |rld.Hello World.|
01b0  48 65 6c 6c 6f 20 57 6f  72 6c 64 00 48 65 6c 6c  |Hello World.Hell|
01c0  6f 20 57 6f 72 6c 64 00  48 65 6c 6c 6f 20 57 6f  |o World.Hello Wo|
01d0  72 6c 64 00 48 65 6c 6c  6f 20 57 6f 72 6c 64 00  |rld.Hello World.|
01e0  48 65 6c 6c 6f 20 57 6f  72 6c 64 00 48 65 6c 6c  |Hello World.Hell|
01f0  6f 20 57 6f 72 6c 64 00  48 65 6c 6c 6f 20 57 6f  |o World.Hello Wo|
avrdude> read eeprom
Reading | ################################################## | 100% 0.00 s 
0000  65 65 70 20 00 65 65 70  20 00 65 65 70 20 00 65  |eep .eep .eep .e|
0010  65 70 20 00 65 65 70 20  00 65 65 70 20 00 65 65  |ep .eep .eep .ee|
0020  70 20 00 65 65 70 20 00  65 65 70 20 00 65 65 70  |p .eep .eep .eep|
0030  20 00 65 65 70 20 00 65  65 70 20 00 65 65 70 20  | .eep .eep .eep |
0040  00 65 65 70 20 00 65 65  70 20 00 65 65 70 20 00  |.eep .eep .eep .|
0050  65 65 70 20 00 65 65 70  20 00 65 65 70 20 00 65  |eep .eep .eep .e|
0060  65 70 20 00 65 65 70 20  00 65 65 70 20 00 65 65  |ep .eep .eep .ee|
0070  70 20 00 65 65 70 20 00  65 65 70 20 00 65 65 70  |p .eep .eep .eep|
0080  20 00 65 65 70 20 00 65  65 70 20 00 65 65 70 20  | .eep .eep .eep |
0090  00 65 65 70 20 00 65 65  70 20 00 65 65 70 20 00  |.eep .eep .eep .|
00a0  65 65 70 20 00 65 65 70  20 00 65 65 70 20 00 65  |eep .eep .eep .e|
00b0  65 70 20 00 65 65 70 20  00 65 65 70 20 00 65 65  |ep .eep .eep .ee|
00c0  70 20 00 65 65 70 20 00  65 65 70 20 00 65 65 70  |p .eep .eep .eep|
00d0  20 00 65 65 70 20 00 65  65 70 20 00 65 65 70 20  | .eep .eep .eep |
00e0  00 65 65 70 20 00 65 65  70 20 00 65 65 70 20 00  |.eep .eep .eep .|
00f0  65 65 70 20 00 65 65 70  20 00 65 65 70 20 00 65  |eep .eep .eep .e|
avrdude> erase
erasing chip (but not EEPROM) ...
avrdude> read eeprom 0 0x100
Reading | ################################################## | 100% 0.00 s 
0000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0020  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0030  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0040  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0050  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0060  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0070  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0080  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
0090  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00a0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00b0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00c0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00d0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00e0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
avrdude> c
config wdtperiod=t_off # 0
config wdtwindow=t_off # 0
config bodsleep=bod_disabled # 0
config bodactive=bod_disabled # 0
config bodsampfreq=bod_128hz # 0
config bodlevel=bod_1v9 # 0
config clksel=oschf # 0
config eesave=eex_preserved # 1
config browsave=0 # 0
config rstpincfg=gpio # 0
config updipincfg=updi # 1
config crcsel=crc16 # 0
config crcsrc=nocrc # 3
config sut=sut_0ms # 0
config usbsink=enable # 1
config codesize=0 # 0
config bootsize=0 # 0
config nvmlevel=basic # 3
config nvmkey=notact # 0x0000
config key=nolock # 0x5cc5c55c
```

File: src/term.c (hunk 1 above)
Comment (dl8dtl) [reply src/term.c:971]: > But what _do_ these parts do on a CE? Erase EEPROM?

Sure. That's the secure default that prevents leaking information if you had locked the device.

Comment (dl8dtl) [reply src/term.c:971]: I'm surprised about some of the missing ATDF files.

Comment (dl8dtl) [reply src/term.c:971]: > what the bootloader does with EEPROM on a CE is anyone's guess

It's not even sure the bootloader does anything at all when instructed to erase "the chip".

**Comment by stefanrueger** [comment]
>  If I exit Avrdude after setting the eesave fuse to 1 and then starting Avrdude again, the EEPROM is preserved on the AVR-DU.

Ahhhh. It's b/c fuse settings are read by the device only on reset! It is important that fuse settings do not take immediate effect, so you can finish a programming session when you change something like clock source or lock. Sooooo, your test has uncovered that it's *neigh impossible* for AVRDUDE to know whether or not a chip erase will save/preserve EEPROM as it cannot know whether the current fuse setting is in effect.

**Comment by stefanrueger** [comment]
> cannot know whether the current fuse setting is in effect

Soooo, should we say erasing chip (*probably* without EEPROM) ... and erasing chip (*probably* including EEPROM)? We could read the eesave configuration at AVRDUDE startup and stash that away to be consulted on terminal erase, *but* AVRDUDE cannot know whether the user changed fuses *and* pressed the target board's reset during terminal use (hehe, they could have transplanted a whole different part...).

The most conservative would be to read fuses on AVRDUDE startup and still say *probably* here.

Or we could give up and not try to augur what the part does on a CE? I now ever so slightly favour this (less code) also for another reason: `factory reset`  *first* resets fuses to factory setting, and then calls the terminal command `erase` as some parts need a CE for the lock bits to be reset. Now, it would look odd if a factory reset were to say erasing chip (*probably* including EEPROM) ... as it defo erases EEPROM. Could correct this too but that's more code.

Anyway, what do people think?

**Comment by dl8dtl** [comment]
Indeed, it's getting complex.

After the chip erase, we could read back the EEPROM to see whether it's empty, but that might take a lot of time, for little added value – if someone has configured the EESAVE fuse, he's hopefully going to know that already anyway.

**Comment by stefanrueger** [comment]
Correct, reading EEPROM would help. But even if we read before and after the CE it could well be that both times the EEPROM is empty, so we *still* wouldn't know whether or not CE has erased EEPROM. The more I think about it the more I think it's best to say what the code does, ie, `performing chip erase ...`, rather than speculating what the chip does.

Fun fact: the terminal cache also needs to update its state after a CE, so what it does is to go through the cached EEPROM copies of pages that were read earlier; it then reads from the device a single page that ought to be non-empty if CE preserved EEPROM. If that's empty after a CE, the cache decides that CE erased EEPROM and updates the cache accordingly.

**Comment by stefanrueger** [comment]
Following tests and discussion this PR no longer seeks to clarify what a CE does b/c AVRDUDE cannot be certain whether the fuses are currently in effect or might come into effect only on next reset. But the discussion has informed the development of a much better dryrun CE emulation, which is now offered instead.

**Comment by stefanrueger** [comment]
And this PR will make AVRDUDE aware that a CE actually erases bootrow...

**Pull request merged**